### PR TITLE
Reduce unnessecary DB query for subject ID

### PR DESCRIPTION
**Pull request by alistair3149**
These commits are originally from the WeirdGloop's fork.

## Summary by CodeRabbit

- **New Features**
	- Enhanced logic for retrieving property values, improving efficiency in handling sort keys and subject IDs.
	- Added filtering for subjects based on data item type to ensure accurate ID retrieval.

- **Bug Fixes**
	- Optimized ID retrieval process to reduce unnecessary database calls when a valid subject ID is available.

- **Documentation**
	- Updated comments to clarify the rationale behind new ID retrieval logic.

### CODE REVIEW

File: src/SQLStore/EntityStore/PrefetchCache.php [hunk 1]
```diff
@@ -146,13 +146,17 @@ public function prefetch( array $subjects, DIProperty $property, RequestOptions
...
+		// there is no need to do a DB query for the ID.
+		$sid = $subject->getId() !== 0
+			? $subject->getId()
+			: $this->store->getObjectIds()->getSMWPageID(
+				$subject->getDBkey(),
+				$subject->getNamespace(),
+				$subject->getInterwiki(),
+				$subject->getSubobjectName(),
+				true
+			);
```
Comment (JeroenDeDauw) [inline src/SQLStore/EntityStore/PrefetchCache.php:159]: This would be much more readable in a dedicated function with a guard clause

**Comment by JeroenDeDauw** [comment]
> WeirdGloop's fork.

What is this fork I keep hearing about, and why does it have all the nice things without them having ended up here already?

**Comment by alistair3149** [comment]
> > WeirdGloop's fork.
> 
> What is this fork I keep hearing about, and why does it have all the nice things without them having ended up here already?

This the the fork: https://github.com/weirdgloop/mediawiki-extensions-SemanticMediaWiki

WG runs several official gaming wikis such as the [RuneScape Wikis](https://runescape.wiki), [Minecraft Wikis](https://minecraft.wiki), and [League of Legends Wiki](https://wiki.leagueoflegends.com/en-us).

I'm not affiliated with them so I can't speak for them. I'm only backporting some of the patches that they did when it is applicable and that I understand the tweak. Some of the tweaks are highly specific to their setup so it might not be useful for the others.

If anyone is interested in reaching out directly, the `#wiki-tech` in their Discord would be the best place: https://discord.gg/runescapewiki

File: src/SQLStore/EntityStore/PrefetchCache.php (hunk 1 above)
Comment (alistair3149) [reply src/SQLStore/EntityStore/PrefetchCache.php:159]: Will follow up with it in a PR

**Pull request merged**